tiledb/api/src/query/write/strategy.rs

```diff
@@ -976,19 +1162,63 @@ mod tests {
                     None
                 };
 
-                array = write_query
+                let _ = write_query
                     .finalize()
                     .expect("Error finalizing write query");
 
                 write_ranges
             };
 
-            array = Array::open(ctx, array.uri(), Mode::Read).unwrap();
+            if write.cells().is_empty() {
+                // in this case, writing and finalizing does not create a new fragment
+                // TODO
+                continue;
+            }
 
             /* NB: results are not read back in a defined order, so we must sort and compare */
 
-            /* first, read back what we just wrote */
+            let mut array = ArrayOpener::new(ctx, &uri, Mode::Read)
+                .unwrap()
+                .open()
+                .unwrap();
+
+            /*
+             * First check fragment - its domain should match what we just wrote, and we need the
+             * timestamp so we can read back only this fragment
+             */
+            let [timestamp_min, timestamp_max] = {
+                let fi = array.fragment_info().unwrap();
+                let nf = fi.num_fragments().unwrap();
+                assert!(nf > 0);
+
+                let this_fragment = fi.get_fragment(nf - 1).unwrap();
+
+                if let Some(write_domain) = write.domain() {
+                    let nonempty_domain =
+                        this_fragment.non_empty_domain().unwrap().untyped();
+                    assert_eq!(write_domain, nonempty_domain);
+                } else {
+                    // most recent fragment should be empty,
+                    // what does that look like if no data was written?
+                }
+
+                this_fragment.timestamp_range().unwrap()
+            };
+
+            /*
+             * Then re-open the array to read back what we just wrote
+             * into the most recent fragment only
+             */
             {
+                array = array
+                    .refresh()
+                    .start_timestamp(timestamp_min)
```

Comment:
Did I miss a sleep in here or a place where we set these timestamps for writes? There's an issue that folks run into (mostly while testing) where its possible to write two fragments within the same millisecond. If that happens you'll not be able to discriminate between those two fragments.

The two general approaches I've seen for handling this particular issue in tests is to either manually specify the timestamps on write, or insert a short 1-2ms sleep.

Reply:
Hm, there's a few assumptions built in then:
1) the fragment list is sorted on timestamp, so that `fi.get_fragment(num_fragments - 1)` always gives a handle to the one we just wrote
2) things are slow enough that the previous fragment's `timestamp_min` is less than the most recent's `timestamp_max`

Inserting a sleep seems safe though I'd be quite surprised if we hit #2 since each loop of the test is quite hefty